guest-owner-proxy/guest-owner-proxy.py

```diff
@@ -0,0 +1,254 @@
+    except Exception as e:
+        print("Failed to load keyfile: {}".format(e.msg))
+
+
+    #makedirs(certs_path, exist_ok = True)
+
+    # this demo implementation isn't designed for parallelism.
+    server = grpc.server(futures.ThreadPoolExecutor(max_workers=1))
+    pre_attestation_pb2_grpc.add_SetupServicer_to_server(SetupService(), server)
+
+    server.add_insecure_port("[::]:{}".format(options.grpc_port))
+    server.start()
+    server.wait_for_termination()
+
+
+class SetupService(pre_attestation_pb2_grpc.SetupServicer):
+    def __init__(self):
+        self.connection_id = 1
+
+    def GetLaunchBundle(self, request, context):
+        print("Servicing Launch Bundle Request")
+
+        # make dir for this conection
+        connection_certs_path = path.join(certs_path, "connection{}".format(self.connection_id))
+        makedirs(connection_certs_path, exist_ok=True)
+
+        # save pdh to file
+        with open(path.join(connection_certs_path,"pdh.cert"),"wb") as f:
+            f.write(request.PlatformPublicKey)
+
+        # generate launch blob
+        # use sevtool for now. might switch in the future
+        cmd = "sudo {} --ofolder {} --generate_launch_blob {}". \
+                format(sevtool_path, connection_certs_path, request.Policy)
+        subprocess.run(cmd.split())
+
+        # read in the launch blob
+        with open(path.join(connection_certs_path, "launch_blob.bin"), "rb") as f:
+            launch_blob = f.read()
+
+        # read in the guest owner public key
+        with open(path.join(connection_certs_path, "godh.cert"), "rb") as f:
+            godh = f.read()
+
+        self.connection_id += 1
+
+        return BundleResponse(GuestOwnerPublicKey = godh, \
+                LaunchBlob = launch_blob, \
+                ConnectionId = connection_id)
+
+
+    # TODO: make into smaller functions
+    def GetLaunchSecret(self, request, context) :
+        print("Serving Launch Secret Request")
+
+        # check each of the parameters against the keyset
+        if not request.KeysetId in keysets:
+            context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+            context.set_details('KEYSET INVALID')
+            return SecretReponse()
+
+        keyset = keysets[request.KeysetId]
+
+        # policy
+        if not request.Policy in keyset['allowed-policies']:
+            context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+            context.set_details('POLICY INVALID')
+            return SecretReponse()
+
+        # api
+        if not request.ApiMajor >= keyset['min-api-major']:
+            context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+            context.set_details('API MAJOR VERSION INVALID')
+            return SecretReponse()
+
+        if not request.ApiMinor >= keyset['min-api-minor']:
+            context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+            context.set_details('API MINOR VERSION INVALID')
+            return SecretReponse()
```

Comment:
Not sure the check of major+minor is correct:

Assume my keyset says: `min-api-major=15, min-api-minor=8` . This means I should accepts guests equal or newer than 15.0008 . Now consider a request has `major=17 minor=3` -- the major is bigger than 15 but the minor is smaller than 8. But I think this should still be considered an allowed API version.

One solution is to check minor version only when `request.ApiMajor==keyset['min-api-major']` (major is equal, the minor check will decide).

Reply:
Yeah. I think what you really want is to treat the major and minor like one number. As in major=15 and minor=14 becomes 15.14. Then you can just compare two versions directly. I couldn't think of a clean way of doing this at the time. I considered dividing the minor by 100 and adding it to the major, but that seemed kind of weird.